functions/functions.go

```diff
@@ -0,0 +1,187 @@
+	Minima() []Minimum
+}
+
+// Minimum represents information about an optimal location of a function.
+type Minimum struct {
+	// X is the location of the minimum. X may not be nil.
+	X []float64
+	// F is the value of the objective function at X.
+	F float64
+	// Global indicates whether the minimum is local or global.
+	Global bool
+}
+
+type FuncTest struct {
+	X []float64
+
+	WantF    float64
+	WantGrad []float64 // If nil, gradient is not evaluated.
+}
+
+// TODO: Decide and implement an exported testing function:
+// func Test(f Function, ??? ) ??? {
+// }
+
+const (
+	defaultTol     = 1e-8
+	defaultGradTol = 1e-5
+)
+
+func testFunction(f Function, tests []FuncTest, t *testing.T) {
+	// Check that the function can evaluate itself (and its gradient) correctly.
+	for i, test := range tests {
+		F := f.F(test.X)
+
+		// Check that the function value is as expected.
+		if math.Abs(F-test.WantF) > defaultTol {
+			t.Errorf("Test #%d: function value given by F() is incorrect. Want: %v, Got: %v",
+				i, test.WantF, F)
+		}
+
+		if test.WantGrad == nil {
+			continue
+		}
+
+		// Evaluate the finite difference gradient.
+		fdGrad := make([]float64, len(test.WantGrad))
+		fd.Gradient(fdGrad, func(x []float64) float64 {
+			return f.F(x)
+		}, test.X, nil)
+
+		// Check that the finite difference and expected gradients match.
+		if !floats.EqualApprox(fdGrad, test.WantGrad, defaultGradTol) {
+			dist := floats.Distance(fdGrad, test.WantGrad, math.Inf(1))
+			t.Errorf("Test #%d: numerical and expected gradient do not match. |fdGrad - WantGrad|_∞ = %v",
+				i, dist)
+		}
+
+		if fGradient, isGradient := f.(Gradient); isGradient {
+			grad := make([]float64, len(test.WantGrad))
+			fGradient.Df(test.X, grad)
+
+			// If the function is a Gradient, check that it computes the gradient correctly.
+			if !floats.EqualApprox(grad, test.WantGrad, defaultGradTol) {
+				dist := floats.Distance(grad, test.WantGrad, math.Inf(1))
+				t.Errorf("Test #%d: gradient given by Df() is incorrect. |grad - WantGrad|_∞ = %v",
+					i, dist)
+			}
+		}
+
+		fFunctionGradient, isFunctionGradient := f.(FunctionGradient)
+		if isFunctionGradient {
+			grad := make([]float64, len(test.WantGrad))
+			F := fFunctionGradient.FDf(test.X, grad)
+
+			// If the function is a FunctionGradient, check that it computes the value correctly.
+			if math.Abs(F-test.WantF) > defaultTol {
+				t.Errorf("Test #%d: function value given by FDf() is incorrect. Want: %v, Got: %v",
+					i, test.WantF, F)
+			}
+
+			// If the function is a FunctionGradient, check that it computes the gradient correctly.
+			if !floats.EqualApprox(grad, test.WantGrad, defaultGradTol) {
+				dist := floats.Distance(grad, test.WantGrad, math.Inf(1))
+				t.Errorf("Test #%d: gradient given by FDf() is incorrect. |grad - WantGrad|_∞ = %v",
+					i, dist)
+			}
+		}
+	}
+
+	fMinima, isMinimumer := f.(Minimumer)
+	if !isMinimumer {
+		return
+	}
+
+	// Check that the listed minima are correct.
```

Comment:
Can we just append the minima to the fd test? Seems like it would reduce all this code duplication.